### PR TITLE
fix(issue-authoring): recognize more prose-dependency edge cases

### LINKED ISSUE #1472: prose-dependency: remaining reference-shape and list-scoping gaps
## Background

#1468 (closed by #1469) implemented exactly the three reference
shapes its own `## Proposed change` named: trailing Markdown-link
content (fragment/slash/title), and local `owner/repo#N` shorthand.
Its plan explicitly deferred four other findings rather than chase
them in the same PR — the same reasoning #1468 itself used when it
was spun out of #1399. This issue consolidates those four so they
have an open tracker instead of living only in closed-issue/merged-PR
comments.

Three were recorded as addenda on #1468 by its author, before this
PR started:

1. **Reference-style Markdown links.** `Before starting, [PR #1391][upstream]
   must land.` with a separate `[upstream]: https://github.com/acme/other-repo/pull/1391`
   definition elsewhere is not consumed by the inline-link
   alternative, so the label's bare `#1391` is picked up by the
   bare-`#` alternative and flagged even when the definition targets
   another repository. Same root category as #1468's trailing-content
   items (Markdown link recognition), different concrete syntax.
2. **Nested list items lose parent scope.** `splitIntoListItemBlocks`
   (added to fix the tight-list sentence-conflation false positive)
   splits at every list-item marker, including nested/child markers.
   A parent bullet stating the coordination language
   (`- Before starting this work:`) and a nested child bullet carrying
   the reference (`  - PR #1391 must land`) end up in different
   blocks, so the check never sees the keyword and the reference
   together and misses a genuine unencoded prose-only dependency. This
   is a false negative in the opposite direction from what the
   original list-item fix targeted, and it is about sentence/block
   scoping rather than reference-shape recognition — a different
   underlying mechanism from items 1, 3, and 4 here, but tracked
   together since it surfaced in the same review cycle and shares the
   same "this check's edge cases are open-ended" cause.
3. **Empty-string `currentRepo` is treated as "known".**
   `currentRepo !== undefined` is true for `""`, so `${owner}/${repo}`
   (never empty) can never equal it — every full-URL/cross-repo-shorthand
   match is silently treated as cross-repo and skipped. Reachable via
   an explicit `--current-repo ''` or an environment where
   `$GITHUB_REPOSITORY` resolves to an empty string. Fix direction:
   trim and normalize an empty (or whitespace-only) `currentRepo` to
   `undefined` before it reaches `checkProseOnlyDependency`.

One more surfaced during #1469's own review (Codex), on the same
title-quoting feature #1469 added:

4. **Escaped quotes inside a Markdown link title break the title
   sub-pattern.** `[PR #12](https://github.com/other/repo/pull/12
   "reviewed \"API\"")` has an escaped `"` inside the quoted title;
   the title sub-pattern (`"[^"\n]*"`) stops at the first unescaped
   `"`, so it doesn't reach the real closing `"`— the match then falls
   through the same way #1468's own trailing-content items did,
   re-leaking the label's bare `#12` to the bare-`#` alternative for a
   cross-repo link. Same root category as item 1 above and #1468's
   items 1-2 (Markdown link recognition), narrower: only the escaping
   convention inside an already-recognized title.

This check is documented as advisory-only and deliberately tolerant of
false positives/negatives (it never flips `passed` or the exit code),
so shipping #1468/#1469 without these four is consistent with that
philosophy, the same way #1468 said of leaving these out of #1399.

Refs #1468

## Proposed change

Extend `ISSUE_OR_PR_REFERENCE_PATTERN` / `checkProseOnlyDependency` /
`splitIntoListItemBlocks` in `src/scripts/audit-authored-issue.mts`:

- Recognize reference-style Markdown links (`[text][ref]` plus a
  `[ref]: url` definition) as a Markdown-link match, applying the same
  `currentRepo` comparison already used for the inline form.
- Allow an escaped quote (`\"` / `\'`) inside a Markdown link title
  without ending the title sub-pattern early, without over-matching
  past the link's real closing paren.
- Normalize an empty or whitespace-only `currentRepo` to `undefined`
  before it reaches `checkProseOnlyDependency`, so it is treated as
  "unknown" rather than "known and non-matching".
- Rework `splitIntoListItemBlocks` (or the scoping it feeds) so a
  nested/child list item stays associated with its parent bullet's
  text for this check, without reintroducing the tight-list
  sentence-conflation false positive the current splitting was added
  to fix.

The last item is a materially different kind of change from the other
three (block-scoping logic vs. pattern/normalization extension) and
carries more regression risk against the existing list-item handling.
A future B2 plan for this issue may split it into its own issue rather
than implement all four together, mirroring how #1468 itself was
narrowed out of #1399 — this issue's job is to give all four an open
tracker, not to pre-commit to one PR shape.

## Acceptance criteria

- A reference-style Markdown link (`[text][ref]` + separate `[ref]: url`
  definition) is recognized like the existing inline-link alternative,
  for both local and cross-repo targets.
- A Markdown link title containing an escaped quote matching the
  title's own quote style is recognized as part of the title, without
  swallowing content past the link's real closing paren.
- An empty-string or whitespace-only `--current-repo` /
  `$GITHUB_REPOSITORY` is treated identically to it being unset.
- A nested/child list item carrying an issue/PR reference is evaluated
  together with its parent bullet's coordination-language text, without
  regressing the tight-list sentence-conflation fix already in place.
- New regression tests cover each of the four items individually and
  in combination with an existing `Blocked by`/`Depends on`/task-list
  marker for the same number.
- `skills/issue-authoring/references/contract.md` and
  `docs/issue-authoring-skill.md` describe whichever of the four shapes
  land; `.claude/skills/issue-authoring/references/contract.md` matches
  after `node scripts/sync-docs.mjs --apply`, and
  `node scripts/audit-docs.mjs --check` passes.
- `pnpm run build`, `pnpm exec tsc --noEmit`, and the full test suite
  pass.

## Candidate files

- `src/scripts/audit-authored-issue.mts` (source)
- `scripts/audit-authored-issue.mjs` (generated mirror)
- `tests/audit-authored-issue.test.mts`
- `skills/issue-authoring/references/contract.md` (canonical) +
  `.claude/skills/issue-authoring/references/contract.md` (generated)
- `docs/issue-authoring-skill.md`

---

_Autopilot suitability: 3 / 5 -- higher is more autopilot-suitable;
below the configured floor is human-oriented. Set one point below
#1468's 4: item 2 (nested-list scoping) is a block-scoping redesign
rather than a pattern extension, and a future B2 plan may need to
split it out (see `## Proposed change`)._

_Effort: S -- author-estimated size; a soft autopilot selection
tie-breaker only._

### PR DESCRIPTION
**Pull request by kurone-kito**
# Summary

Implements the four prose-dependency edge cases consolidated by #1472
(after #1468/#1469's own review, and #1399's before that, kept finding
related gaps in this same advisory check):

- **Reference-style Markdown links.** `[text][ref]` + a separate
  `[ref]: target` definition is now resolved to the equivalent
  inline-link shape before matching, so the existing Markdown-link
  alternative (and its `currentRepo` comparison) recognizes it with no
  duplicated logic. A `ref` with no matching definition falls through
  to the bare-`#` alternative like any other unrecognized shape.
- **Nested list items losing parent scope.** `splitIntoListItemBlocks`
  now compares each marker line's indentation against the *most
  recently seen* marker's indentation instead of splitting at every
  marker unconditionally: a strictly deeper marker (nested/child) stays
  merged with the running block, while same-depth markers (nested or
  top-level) still start a new block — preserving the existing
  tight-list sentence-conflation fix this function exists for.
- **Empty-string `currentRepo`.** `currentRepo !== undefined` was true
  for `''`, so every full-URL/shorthand match was silently treated as
  cross-repo and excluded. An empty or whitespace-only value is now
  normalized to `undefined` before it reaches the check.
- **Escaped quotes in a Markdown link title.** The quoted-title
  sub-pattern now tolerates a backslash-escaped quote matching its own
  delimiter (`\"` / `\'`) as content instead of treating it as the
  closing quote, which previously failed the whole Markdown-link
  alternative and re-leaked the label's bare `#N` to the bare-`#`
  alternative.

This check remains advisory-only (never flips `passed` or the exit
code).

## Linked issue

Closes #1472

## Follow-up issues (if any)

- [x] none

## Background / rationale (if material)

Per the issue's own scope note, only these four items are implemented
in this PR. Item 2 (nested-list scoping) carries the most regression
risk — it was validated against the existing tight-list sentence-
conflation regression test plus new sibling-still-separates and
multi-child-residual checks (see test file). It was not split into a
separate issue since this task requires landing all four together in
one PR; a documented, accepted residual limitation remains for a parent
bullet with two or more nested children (only the first child stays
merged with the parent — no worse than today's behavior, which merges
none).

## IDD impact

- [ ] Instruction files changed
- [ ] Template files changed
- [ ] Helper scripts changed
- [ ] Config schema changed
- [ ] Security / credential / merge behavior changed

## Verification

- [x] `pnpm lint` (typecheck, `build:check`, biome, dprint, `node --test`,
      workshop-integrity, cspell, markdownlint — equivalent commands run
      individually, see below)
- [x] `pnpm test` — `node --test tests/audit-authored-issue.test.mts`:
      93/93 pass (78 pre-existing + 15 new regression tests covering
      each of the 4 items individually, cross-repo/unknown-currentRepo
      polarity, and combination with an existing `Blocked by` encoding).
      Full suite `node --test tests/*.test.mts`: 2294/2297 pass; the 3
      failures are all in `tests/branch-conflict-state.test.mts`
      (`classifyBranchConflictState: ...`), reproduced independently in
      complete isolation — a pre-existing fixture GPG-pinentry-timeout
      issue in this local environment, unrelated to this diff.
- [x] `node scripts/audit-docs.mjs --check`
- [x] `pnpm run docs:sync:check`
- [x] `node scripts/idd-doctor.mjs` (pre-existing, unrelated warnings only)

## Safety

- [x] No secret-bearing examples
- [x] No private repo names / URLs
- [x] Merge policy impact reviewed
- [x] Context budget impact reviewed

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved detection of issue and pull request references in Markdown, including reference-style links and link titles with escaped quotes.
  * Corrected local vs cross-repository reference handling when repository context is missing or whitespace-only.
  * Improved advisory sentence-scoping for nested list items to prevent incorrect warnings.

* **Documentation**
  * Expanded guidance for `prose-dependency` advisory behavior, including reference matching rules, `--current-repo` handling, and nested list scope limitations.

* **Tests**
  * Added coverage for reference-style link resolution, escaping edge cases, repository context trimming/unknown behavior, and nested list coordination scenarios.

### CODE REVIEW

**Comment by kurone-kito** [comment]
CodeQL flagged 4 high-severity "inefficient regular expression" alerts on the first push (exponential backtracking in the escaped-quote title sub-pattern this PR added for item 4). Verified the vulnerability empirically (a ~30-char adversarial title took several seconds; grew exponentially with input length) and fixed it in 7ec0be74 by excluding the backslash from the character-class alternative so the two alternatives never overlap — the standard non-overlapping idiom for a backslash-escaped quoted string. Added a regression test asserting linear-time evaluation on a 5000-repetition adversarial input. Re-verified the fix at ~10,000 characters resolves in under 1ms.

File: src/scripts/audit-authored-issue.mts [hunk 1]
```diff
@@ -206,18 +210,59 @@ const PROSE_DEPENDENCY_KEYWORDS = [
...
+// TDZ-avoidance placement rationale as LIST_ITEM_MARKER_PATTERN above.
+const REFERENCE_STYLE_LINK_USAGE_PATTERN = /\[([^\]\n]*)\]\[([^\]\n]+)\]/g;
+
+// Matches a Markdown link reference definition line: optionally indented
+// (up to 3 spaces, per CommonMark), `[label]: target`, with the target
+// read up to the first whitespace. An optional title on the same
+// definition line (e.g. `[ref]: <url> "title"`) is intentionally not
+// captured — only the destination matters for resolving a reference-style
+// link to a GitHub issue/PR URL.
+const LINK_REFERENCE_DEFINITION_PATTERN = /^ {0,3}\[([^\]\n]+)\]:\s*(\S+)/gm;
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/audit-authored-issue.mts:282]: **Accepted** — confirmed: reference-style link definitions using CommonMark's angle-bracket destination form (`[ref]: <https://...>`) were stored verbatim, producing a rewritten `[text](<url>)` that the Markdown-link alternative does not recognize, re-leaking the label's bare `#N`. Fixed in ea2f2803 by stripping a single surrounding pair of angle brackets before storing the destination; added a regression test.

File: src/scripts/audit-authored-issue.mts [hunk 2]
```diff
@@ -710,6 +784,44 @@ function checkProseOnlyDependency(
...
+ * reference definitions.
+ */
+function resolveReferenceStyleLinks(text: string): string {
+  const definitions = new Map<string, string>();
+  for (const match of text.matchAll(LINK_REFERENCE_DEFINITION_PATTERN)) {
+    const key = normalizeLinkReferenceLabel(match[1]);
+    if (!definitions.has(key)) {
+      definitions.set(key, match[2]);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/audit-authored-issue.mts:845]: **Accepted** — confirmed: reference-style link definitions using CommonMark's angle-bracket destination form (`[ref]: <https://...>`) were stored verbatim, producing a rewritten `[text](<url>)` that the Markdown-link alternative does not recognize, re-leaking the label's bare `#N`. Fixed in ea2f2803 by stripping a single surrounding pair of angle brackets before storing the destination; added a regression test.

File: src/scripts/audit-authored-issue.mts [hunk 3]
```diff
@@ -710,6 +801,44 @@ function checkProseOnlyDependency(
...
+  if (definitions.size === 0) {
+    return text;
+  }
+  return text.replace(
+    REFERENCE_STYLE_LINK_USAGE_PATTERN,
+    (whole: string, label: string, ref: string) => {
+      const target = definitions.get(normalizeLinkReferenceLabel(ref));
+      return target === undefined ? whole : `[${label}](${target})`;
+    },
+  );
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/audit-authored-issue.mts:855]: **Accepted** — confirmed: reference-style link definitions using CommonMark's angle-bracket destination form (`[ref]: <https://...>`) were stored verbatim, producing a rewritten `[text](<url>)` that the Markdown-link alternative does not recognize, re-leaking the label's bare `#N`. Fixed in ea2f2803 by stripping a single surrounding pair of angle brackets before storing the destination; added a regression test.

File: src/scripts/audit-authored-issue.mts [hunk 4]
```diff
@@ -710,6 +784,44 @@ function checkProseOnlyDependency(
...
+    }
+  }
+  if (definitions.size === 0) {
+    return text;
+  }
+  return text.replace(
+    REFERENCE_STYLE_LINK_USAGE_PATTERN,
+    (whole: string, label: string, ref: string) => {
+      const target = definitions.get(normalizeLinkReferenceLabel(ref));
+      return target === undefined ? whole : `[${label}](${target})`;
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/audit-authored-issue.mts:853]: **Accepted** — confirmed: reference-style link definitions using CommonMark's angle-bracket destination form (`[ref]: <https://...>`) were stored verbatim, producing a rewritten `[text](<url>)` that the Markdown-link alternative does not recognize, re-leaking the label's bare `#N`. Fixed in ea2f2803 by stripping a single surrounding pair of angle brackets before storing the destination; added a regression test. (Thread already auto-resolved.)

File: src/scripts/audit-authored-issue.mts [hunk 5]
```diff
@@ -740,17 +852,46 @@ function splitIntoSentences(paragraph: string): string[] {
...
+      (lastMarkerIndent === null || marker[1].length <= lastMarkerIndent);
+    if (startsNewBlock) {
       blocks.push(current.join('\n'));
       current = [line];
     } else {
       current.push(line);
     }
+    if (marker !== null) {
+      lastMarkerIndent = marker[1].length;
+    }
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/audit-authored-issue.mts:946]: **Rejected** — verified, real gap, but deferred rather than fixed in this PR. This is the known, documented residual limitation from #1472's own design (a parent bullet with two or more nested children only keeps the first child merged with the parent's coordination text; a full fix requires tracking a complete indentation-ancestry stack, a materially larger and higher-regression-risk change than the most-recent-marker comparison this PR implements — the same trade-off #1472's own body anticipated needing a separate issue for, and independently confirmed as acceptable by a critique pass before implementation). #1472 is the third generation of a similar deferral chain (#1399 → #1468 → #1472); per explicit task scope this pattern terminates here rather than expanding further. Filed as a tracked, consolidated follow-up: #1474.

**Comment by kurone-kito** [comment]
**Rejected** — coderabbitai[bot] did not review HEAD ea2f28033e9bcf9b0950b1aa645a28de7c565d86 (rate-limited per its own comment, next review window ~39 minutes out at post time); this is not a completed review.

**Comment by kurone-kito** [comment]
_claude-7300c0e3: critique baseline — IDD automation marker. Do not edit._

File: src/scripts/audit-authored-issue.mts [hunk 6]
```diff
@@ -740,17 +885,46 @@ function splitIntoSentences(paragraph: string): string[] {
...
   const blocks: string[] = [];
   let current: string[] = [];
+  let lastMarkerIndent: number | null = null;
   for (const line of paragraph.split('\n')) {
-    if (LIST_ITEM_MARKER_PATTERN.test(line) && current.length > 0) {
+    const marker = LIST_ITEM_MARKER_PATTERN.exec(line);
+    const startsNewBlock =
+      marker !== null &&
+      current.length > 0 &&
+      (lastMarkerIndent === null || marker[1].length <= lastMarkerIndent);
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/audit-authored-issue.mts]: **Accepted** — confirmed: CommonMark expands a tab to the next column that is a multiple of 4 when it participates in block structure, so a raw `.length` comparison undercounts a tab-indented marker's effective depth, wrongly starting a new block instead of keeping a tab-indented nested child scoped with a space-indented parent. Fixed by comparing expanded column width (`indentColumnWidth`) instead of raw character count; added a regression test.

**Comment by kurone-kito** [comment]
_claude-7300c0e3: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
| Field | Value |
| --- | --- |
| Phase | F3 merged |
| Claim | claude-7300c0e3 / c359a07d-a5ea-467f-8950-a3b856f55ce4 |
| Branch | issue/1472-prose-dependency-remaining-reference |
| Last checked | 2026-07-17T19:32:05.818Z |
| Open blockers | none |
| Next action | F4 cleanup then F5 discover |
| Authoritative by | merge commit 7a84fe99b1d8b1bc30d6ed2e479f49de4c3f8f1e, matched head 3103699638f444c93ff2aae995d34a2d1be11879 |
